---
title: "Vignette__Pathway_Enrichment_Analysis"
author: "Christina Hillig"
date: "3/15/2021"
output:
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(eval = TRUE)
```

# Introduction
The following section describes the pathway enrichment analysis of up-regulated genes in the IL17A transcript positive and negative group in Leukocytes in the ST data set. 

# Initialisation
Load required R and Bioconductor packages:
```{r, message=FALSE}
# R packages
rlibs <- c("dplyr", "gtools", "hash", "kableExtra", "knitr", 
           "stringr", "tibble", "xlsx", "hash", "Hmisc")
invisible(lapply(rlibs, require, character.only = TRUE))
# Bioconductor packages
bioclibs <- c("ReactomePA", "pathview",  "enrichplot", "org.Hs.eg.db", "DOSE")
invisible(lapply(bioclibs, require, character.only = TRUE))

getwd()
environment()

# Source R-scripts
source(file.path("..", "..", "r_scripts", "pathway_analysis", 'init.R'))
source(file.path("..", "..", "r_scripts", "pathway_analysis", 'load_data.R'))
source(file.path("..", "..", "r_scripts", "pathway_analysis",'utils.R'))
# R-script to import immune Pathways which shall be highlighted in the publication
source(file.path("..", "..", "r_scripts", "pathway_analysis", 'pathway_lists.R'))
# R-script to plot the Pathway enrichment result
source(file.path("..", "..", "r_scripts", "pathway_analysis", "plot_pathways.R"))
```

Define input directory.
```{r}
## Input 
# Input directory 
# TODO change to relative path
# input.folder <- file.path("..", "..", "input", "pathway_analysis")
input.folder <- file.path("/Users/christina.hillig/PycharmProjects/ST_Immune_publication/Publication_analysis/input/pathway_analysis")
# Date
date <- "2022-05-11"
# Data set 
dataset.type <- 'Leukocytes'
# Sequencing technique
seq.technique <- "ST"
# Comparison
cytokine <- 'IL13'
comparison <- paste(cytokine, '_vs_Others', sep="")
plot_cytokine <- TRUE

# Used design function:
design.function <- "cdr_project_patient_annotation_condition"

# General input directory
input.dir <- get_inputdir(input.folder = input.folder, date.file = date,
                         dataset.type = dataset.type, seq.technique = seq.technique, 
                         comparison = comparison, design.function = design.function,
                         genename=cytokine)

# print(input.dir)
```
Determine cut-parameters to identify differential expressed genes and enriched pathways.
```{r}
# Cut parameter
l2fc.factor <- 1
fdr.value <- 0.05
pval.cut <- 0.05
p.value <- 0.05
minGSSize <- 10
# Multi-test method Benjamini-Hochberg (BH)
multitest.method <-  "BH"
```
```{r}
# Plot Parameters
show_dotplot_categories <- 5

width_img <- 16
height_img <- 12
```
Create output directory.
```{r}
## Output
# output.folder <- file.path("..", "..")
output.folder <- "/Users/christina.hillig/PycharmProjects/ST_Immune_publication/Publication_analysis"
# General output directory
output.dir <- get_outputdir(output.folder = output.folder, dataset.type = dataset.type,
                            seq.technique = seq.technique, genename=cytokine)
# print(output.dir)
```
Load DGE analysis results and prepare dataframe for further analysis steps.
```{r}
# 1. Get all DGE .csv files in subfolders
all_filenames <- list.files(path = input.dir, pattern = ("*.csv|*.xlsx"), recursive = TRUE)
# 2. remove metaData from list
dge_filename <- all_filenames[!grepl("metaData*", all_filenames)]
# 3. Load DGE Analysis result file with colnames: 
# "X" "gene_symbol" "gene_name" "entrezid" "pval" "padj" "log2fc" "hkg"
df.dge_res <- load_files(path_name_file = file.path(input.dir,  dge_filename))

# 4. Filter data for duplicates, unwanted columns
df.dge_res <- filter_data(df.data = df.dge_res, signature_gene = cytokine,
                          plot_signaturecytokine=TRUE) 
print.data.frame(df.dge_res[1:10, ], digits = 4)
```

# Pathway enrichment analysis
The pathway enrichemnt analysis included the follwoing steps:
1. Identify significant and background genes 
1. Run PA analysis
1. Plot Save PA analysis results  

In a first step, we convert the gene symbol to entrezID. That is needed to use the later the ReactomePA::enrichPathway function. In best case use always the esemblID.
```{r}
df.dge_res <- rename_genetoentrezid(df.dge_results = df.dge_res)
```


## Get significantly DEx genes and background genes
Now we can start with idetifying differential expressed genes with our manually set cut parameters.
```{r}
# I. Define significant DEx genes and background genes
# I.a) sort genes into groups belonging either to reference or test (control) condition
df.ref_degenes <- get_significantgenes(df.dge_results = df.dge_res, p_value = p.value,
                              lfc_factor = l2fc.factor, op = `>`) 
df.ref <- df.ref_degenes[[1]]
# I. Rank genes based on their fold change
ranked_genes.ref <- do_rank_genes(df.dge_results = df.ref)

df.ctrl_degenes <- get_significantgenes(df.dge_results = df.dge_res, p_value = p.value,
                                       lfc_factor = -l2fc.factor, op = `<`) 
df.ctrl <- df.ctrl_degenes[[1]]

# I. Rank  genes based on their fold change
ranked_genes.ctrl <- do_rank_genes(df.dge_results = df.ctrl)

# I.b) Background genes are all genes from our (sup-) data set
bg_genes <- as.character(df.dge_res$entrezid)
```


## Identify enriched pathways using ReactomePA
```{r}
# II.a) Find enriched Pathways for reference condition
reactome_object.ref <- ReactomePA::enrichPathway(
  gene = df.ref_degenes[[2]], # a vector of entrezID
  universe = bg_genes, organism = 'human', 
  qvalueCutoff = fdr.value, pvalueCutoff = pval.cut,  pAdjustMethod = multitest.method, 
  minGSSize = minGSSize, maxGSSize = 500, readable = T)

# II.b) Find enriched Pathways for control condition
reactome_object.ctrl <- ReactomePA::enrichPathway(
  gene = df.ctrl_degenes[[2]], # a vector of entrezID
  universe = bg_genes, organism = 'human', 
  qvalueCutoff = fdr.value, pvalueCutoff = pval.cut, pAdjustMethod = multitest.method, 
  minGSSize = minGSSize, maxGSSize = 500, readable = T)
```

In order to plot gene names instead of entrezIDs they have to be converted. For this use the function "*DOSE::setReadable*" to convert entrezIDs to gene symbol.
```{r}
################### ---> convert gene ID to Symbol <--- ################### 
reactome.ctrl <- setreadable_pa(paenrich_object = reactome_object.ctrl)
print("Pathways associated with cytokine-negative group")
print.data.frame(reactome.ctrl[1:3, ], digits = 4)
```
```{r}
reactome.ref <- setreadable_pa(paenrich_object = reactome_object.ref)
print("Pathways associated with cytokine-positive group")
print.data.frame(reactome.ref[1:3, ], digits = 4)
```


## Save and Plot Pathwyas
```{r}
#############################################################################
################### ---> Save results to csv file <--- ###################### 
#############################################################################
# Attention: 
# ctrl (= negative log2FC) and ref (= positive log2FC) are switched for Immune publication
save_enrichobject_as_csv(paenrich_object = reactome.ctrl, condition = 'Cytoneg', 
                         pa_database = 'REACTOME', output_path = output.dir) 
save_enrichobject_as_csv(paenrich_object = reactome.ref, condition = 'Cytopos', 
                         pa_database = 'REACTOME', output_path = output.dir) 
```

Visualise Pathways in a cnet- and dotplot.
```{r}
# III.a) Plot variables
# select pathways or Enriched gene sets manually
publication_pas <- pathwaysofinterest()
if (seq.technique == 'SC') 
{
  pas_publication <- grep(paste('sc', cytokine, sep = "_"), hash::keys(publication_pas),
                         value = TRUE)
} else 
{
  pas_publication <- grep(paste('st', cytokine, sep = "_"), hash::keys(publication_pas),
                         value = TRUE)
}
show_categories <- publication_pas[[pas_publication]]
```

First, we plot the enriched pathways of the *IL17A*-positive group and save the plots as .pdf.
```{r,  fig.width=width_img, dpi = 300 , fig.height=height_img}
# III.b) Reference Condition
# If a gene is associated with two or more enriched PAs 
# but less than those are shown than this results in a bug 
# ==> the log2FC of that gene is not correctly shown
if (!is.null(nrow(reactome.ref)))
{
  if (nrow(reactome.ref) > 1 & any(show_categories %in% reactome.ref$Description))
  {
    # Cnetplots to visualise enriched pathways
    fig.pathways.REACTOME(reactome_res = reactome.ref,
                          entrezid_log2fc = ranked_genes.ref,
                          showCategories = show_categories, 
                          sig.cytokine = cytokine,
                          output.dir = output.dir,
                          title = "Cytopos_REACTOME_Pathway_Enrichment_Analysis.pdf", 
                          width = width_img, height = height_img)

    # Dotplot to visualise enriched pathways
    fig.pathway.dotplot(pathway_res = reactome.ref,
                        showCategories = show_dotplot_categories,
                        method = 'REACTOME', 
                        output.dir = output.dir,
                        title = "Cytopos_REACTOME_dotplot.pdf", 
                        width = 6, height = 4)
    
    
    # TODO workflow figure ..
    # keep only enriched pathways p.adjust <= 0.1
    df_sebg_nl <- reactome.ref
    df_sebg_nl <- df_sebg_nl[df_sebg_nl$p.adjust <= p.value, ]
    # sort by p.adj
    df_sebg_nl <- df_sebg_nl[order(df_sebg_nl$p.adjust),]
    
    
    df_sebg_nl$GeneRatio <- as.numeric(sapply(df_sebg_nl$GeneRatio, function(x) eval(parse(text=x))))
    
    # create interval for count
    df_sebg_nl$Count <- as.numeric(df_sebg_nl$Count)
    breaks <- seq(min(df_sebg_nl$Count), max(df_sebg_nl$Count), 10)
    df_sebg_nl$Counts <- cut(df_sebg_nl$Count, c(breaks, Inf), labels = breaks, include.lowest = TRUE)
    
    
    # Top 15
    df_sebg_nl <- df_sebg_nl[c(1:15), ]
    
    # sort by GeneRatio
    df_sebg_nl <- df_sebg_nl[order(df_sebg_nl$p.adjust, decreasing = FALSE), ]
    
    # Workaround to show categoires not in aplhabetical order
    df_sebg_nl$Description <- factor(df_sebg_nl$Description, levels=unique(df_sebg_nl$Description))
    df_sebg_nl$Description <- factor(df_sebg_nl$Description, levels=df_sebg_nl$Description)
    # SHow only top pathways
    df_sebg_nl_workflow <- df_sebg_nl[c(1:show_dotplot_categories), ]
    
    pdf(file = file.path(output.dir, "Cytopos_REACTOME_barplot.pdf"), width = 8, height = 6)
    p1 <- ggplot(df_sebg_nl_workflow, aes(Description, GeneRatio, fill = p.adjust)) +
      geom_bar(stat = "identity", position = "dodge") +
      ggplot2::theme_minimal() +
      coord_flip() + theme(text = element_text(size=24), axis.title.y=element_blank()) +
      scale_x_discrete(limits=rev) +
      scale_fill_gradient(low="blue", high="red", trans = 'reverse') 
    print(p1)
    dev.off()
  }
}

```
Second, we visualise the enriched pathways in the *IL17A*-negative group.
```{r}
# # III.c) Control Condition
if (!is.null(nrow(reactome.ctrl)))
{
  if (nrow(reactome.ctrl) > 1 & any(show_categories %in% reactome.ctrl$Description))
  {
    # Cnetplots to visualise enriched pathways
    fig.pathways.REACTOME(reactome_res = reactome.ctrl,
                          entrezid_log2fc = ranked_genes.ctrl,
                          showCategories = show_categories,
                          sig.cytokine = cytokine,
                          width = width_img, height = height_img,
                          output.dir = output.dir, 
                          title = "Cytoneg_REACTOME_Pathway_Enrichment_Analysis.pdf")

    # Dotplot to visualise enriched pathways
    fig.pathway.dotplot(pathway_res = reactome.ctrl,
                        showCategories = show_dotplot_categories,
                        method = 'REACTOME',
                        width = 6, height = 4,
                        output.dir = output.dir, 
                        title = "Cytoneg_REACTOME_dotplot.pdf")
  }
}
```

# References
Cite used Bioconductor packages: 
```{r}
citation("ReactomePA")
citation("DOSE")
citation("org.Hs.eg.db")
citation("enrichplot")
citation("pathview")
```

```{r}
sessionInfo()
```


